packages/headless/src/controllers/commerce/product-listing/headless-product-listing.ts

```diff
@@ -0,0 +1,132 @@
+import {Schema, StringValue} from '@coveo/bueno';
+import {CommerceAPIErrorStatusResponse} from '../../../api/commerce/commerce-api-error-response';
+import {CommerceEngine} from '../../../app/commerce-engine/commerce-engine';
+import {configuration} from '../../../app/common-reducers';
+import {ProductRecommendation} from '../../../commerce.index';
+import {
+  fetchProductListing,
+  setProductListingUrl,
+} from '../../../features/product-listing/v2/product-listing-v2-actions';
+import {productListingV2Reducer as productListing} from '../../../features/product-listing/v2/product-listing-v2-slice';
+import {loadReducerError} from '../../../utils/errors';
+import {validateOptions} from '../../../utils/validate-payload';
+import {
+  buildController,
+  Controller,
+} from '../../controller/headless-controller';
+
+const optionsSchema = new Schema({
+  url: new StringValue({
+    required: true,
+    url: true,
+  }),
+});
+
+export interface ProductListingOptions {
+  /**
+   * The URL used to retrieve the product listing.
+   */
+  url: string;
+}
+
+export interface ProductListingProps {
+  /**
+   * The initial options that should be applied to this `ProductListing` controller.
+   */
+  options?: ProductListingOptions;
+}
+
+/**
+ * The `ProductListing` controller allows the end user to configure and retrieve product listing data.
+ */
+export interface ProductListing extends Controller {
+  /**
+   * Changes the URL used to retrieve the product listing.
+   * @param url - The new URL.
+   */
+  setUrl(url: string): void;
+
+  /**
+   * Fetches the product listing.
+   */
+  refresh(): void;
+
+  /**
+   * A scoped and simplified part of the headless state that is relevant to the `ProductListing` controller.
+   */
+  state: ProductListingState;
+}
+
+export interface ProductListingState {
+  products: ProductRecommendation[];
+  error: CommerceAPIErrorStatusResponse | null;
+  isLoading: boolean;
+  responseId: string;
+  url: string;
+}
+
+export type ProductListingControllerState = ProductListing['state'];
+
+/**
+ * Creates a `ProductListingController` controller instance.
+ *
+ * @param engine - The headless commerce engine.
+ * @param props - The configurable `ProductListingController` properties.
+ * @returns A `ProductListingController` controller instance.
+ */
+export function buildProductListing(
+  engine: CommerceEngine,
+  props: ProductListingProps = {}
+): ProductListing {
+  if (!loadBaseProductListingReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+  const getState = () => engine.state;
+
+  const options = {
+    ...props.options,
+  };
+
+  validateOptions(engine, optionsSchema, options, 'buildProductListing');
+
+  dispatch(
+    setProductListingUrl({
+      url: options.url!,
```

Comment:
Since the controller will simply "not work" if the URL is not defined (ie: we won't even be able to query the API), then the URL should be a required prop on the controller.

ie: Make props and `options` in `ProductListingProps` non optional/mandatory. This will remove the force cast you need to do here.

And then add some validation that there's actually a valid URL (non-empty string, with the bueno package) if an implementer use JavaScript instead of TS: In this case they would not get any compiler validation that some value is mandatory, and would end up with some cryptic error message down the line.

You can see this pattern in some other controller, if you search for the function `validateOptions` in `validate-payload.ts`

We can leave the `setURL` function below, but it should be "impossible" to arrive at this line here in the code with no URL.

Reply:
Good point! I was following the existing product listing controller implementation by having the url be optional but there's no need to do that.

Commit incoming!

Reply:
Actually, by not validating the options we receive, we delay the validation to the `setProductListingUrl` action, which only validates the payload and doesn't throw if the url's invalid.

Don't we want to throw if an invalid url is fed to the controller?

Reply:
The function execution does not stop if the url is empty. This will rightfully log a `SchemaValidationError` error, but the controller will still be created, and be in an unusable state, as the `url` will be the one from `getProductListingInitialState().url` :
```js
const engine = buildCommerceEngine({
  configuration: {
    ...getSampleSearchEngineConfiguration(),
  },
});

const productListing = buildProductListing(engine, {
  options: {
    url: ''
  }
})
console.log(productListing);
```

To prevent this, I think we should have validation on the controller options, and not only in the action. This would completely prevent an invalid controller from being created. 

Am I misunderstanding the use case?

Reply:
My bad, you are right !